### DISCUSSION TITLE
Priority manager

### DISCUSSION THREAD

**Discussion by marcocastignoli** [sourcifyeth/sourcify-metamask-snap#2] [Ideas]
idea by @ligi 

Having all the snap in the same view loading async, and some priority manager that shows at the top snaps with important warnings, other snaps will be shown above collapsed with just a green tick

**Comment by Tbaut** [comment]
I don't think there is any way to have some collapse/expand view here, since MM only allows key/value pairs. Also I'm not 100% certain how the async loading works but I believe that a general loader is displayed as long as any promise from the insights is still pending. Meaning that if one of the insight is taking a long time, then none will be displayed unfortunately.

I guess there should be some experiments made here, but there's not a lot of margin for tinkering unfortunately.

**Reply by marcocastignoli** [reply]
I described the desired behavior, right now we can:
* async is not possibile
* expanding could mean give less insights if everything is ok:
e.g. 
  * if all ok: 
**verified**
sourcify :green_circle: 
  * if not all ok
**verified**
sourcify :yellow_circle: 
**verification warning**
partial verification, do not trust informations extracted from the metadata

**Reply by Tbaut** [reply]
I think this is good. Since we're using TS what do you think about defining an interface for insights to return, such as `InsightResult` below that each snap willing to be integrated should have to comply to.
Something like
```ts
enum Score {
    High = "HIGH",
    Medium = "MEDIUM",
    Low = "LOW"
}

interface DetailedResult {
    score: Score;
    description: string;
}

interface InsightResult {
    overallScore: Score;
    overallTitle: string;
    overallDescritpion: string;
    detailedResults: DetailedResult[]; // this will only be displayed if the overallScore is not high
}
```

and then to have a unified overview, you can decide when calling each insight what to show.

```ts
export const onTransaction: OnTransactionHandler = async ({ transaction }) => {
    const noticeInsight: InsightResult = getNoticeInsight(transaction as Transaction)
    const verificationInsight: InsightResult = getVerificationInsight(transaction as Transaction)

    const insights = await Promise.all([noticeInsight, verificationInsight])
        .then((res: InsightResult[]) => {
            return res.reduce((result, currentInsight) => {
                let detailedResult = {}

                // populate the detailed results if the score isn't the best
                // for users to understand what to be careful about
                if (currentInsight.overallResult !== Score.High) {
                    currentInsight.detailedResults.forEach((detail) => (
                        detailedResult = {
                            ...detailedResult, // this 
                            [detail.score]: detail.description
                        }))
                }

                return {
                    [currentInsight.overallTitle]: `${currentInsight.overallResult}${currentInsight.overallDescritpion}`,
                    ...detailedResult, // this will only be displayed if the overallScore is not high
                    ...result
                }
            }, {})
        })

    return {
        insights,
     } as OnTransactionResponse;
};
```

This way, you don't let each insight decide on what is shown and when, and the overall design of the monolithic snap will be more coherent.
